server/queue/queue-file/src/main/java/org/apache/james/queue/file/FileMailQueue.java

```diff
@@ -395,21 +411,45 @@ public long clear() throws MailQueueException {
      */
     @Override
     public long remove(Type type, String value) throws MailQueueException {
```

Comment:
are we thread-safe ?

Reply:
I **guess** yes we are. Here I see two sources of potential danger for concurrency.

* `keyMappings::remove` - is a bulletproof.
* `indexWriter::deleteDocuments` - Again quote from javadoc

> All given deletes are applied and flushed atomically at the same time.

Maybe I should write tests on it to not **guess**.

Reply:
Oh yes, this one is for sure. The `indexReader` will be not _fsync_-ed if concurrent thread add mail to the queue.